### ISSUE TITLE
New Utility: focus manager + ShellPanel enhancement

### ISSUE THREAD

**Issue by asangma** [Esri/calcite-components#1471]
### Description
Add a focus management utility that allows developers to consistently manage focus "tabbing" from UI triggers into the element that was invoked from that trigger and back to the trigger when the element is closed/destroyed.

Currently, app developers must implement their own solutions when browser focus needs to move in a non-linear fashion. For example, in ShellPanel, when an Action in ActionBar is used to trigger the opening of a Panel, the app developer must manually move browser focus to the close button.

### Initial usage
Also use this utility to add default focus "bouncing" in ShellPanel.
An issue was already created for this purpose: #890 

#### User Stories
As an app developer, I want an easy way to manage focus so that I don't have to write it myself.

### Acceptance Criteria 
* Works asynchronously
* Makes it easy for micro-apps and the macro-app to handshake
* ShellPanel integration follows a11y best practices.

**Comment by julio8a** [comment]
@jcfranco and @asangma, is this issue still valid?

**Comment by macandcheese** [comment]
@asangma @jcfranco - checking on this - is this in scope?

**Comment by macandcheese** [comment]
@benelan @jcfranco - suggest closing this, the `setFocus` method lets developers manage this as they wish already, it's not much to have a monitored panel state and then focus the associated panel's close button when it opens / focus the invoking action button when the panel is dismissed.

**Comment by geospatialem** [comment]
@asangma Does the [`setFocus`](https://developers.arcgis.com/calcite-design-system/components/action-bar) method in the `action-bar` component fulfill the intent?

**Comment by macandcheese** [comment]
I think this can be closed - setting `setFocus()` on a panel's `dismiss-button` back to the invoking action when the panel emits the `onCalcitPanelDismissedChange` event is pretty straightforward to manage.

**Comment by geospatialem** [comment]
Closing per the discussion above, with the `setFocus` method.

**Comment by msamwill** [comment]
Is there a write up on this pattern?  I made all my panels closable and used setfocus in action-bar click event to set focus on open panel.  In the calcitePanelClose event I set focus back to the action-bar item.  However, after closing a panel I could no longer set focus to it when opening it again.  Instead, I had to create a custom close button and simply hide the panel versus closing it.  I'm not sure what I did wrong,  

Something not discussed earlier, if the user navigates outside the open panel via TAB key would we want the tab key to bring them back to the open panel from the corresponding action-bar item?  Would we need to monitor the focusout event of the action-bar and if the action-bar item has an open panel set focus to the open panel?

**Comment by macandcheese** [comment]
Hi @msamwill - you should be able to use the `calcitePanelClose` event to return focus to the action that originally invoked opening the Panel: https://codepen.io/mac_and_cheese/pen/abPdjWx?editors=1000 - let me know if that isn't working for your case.

For the last question, do you mean enabling a focus trap for the user inside the Panel? Or returning the focus to the invoking action when a user "tabs out" of the Panel, instead of the next element in the page after the Shell Panel?

**Comment by msamwill** [comment]
Thank you for responding!  What I've done so far is when you click on an action-bar item it opens the panel and sets the focus to the panel.  When you close the panel it returns focus to the action-bar item.  I will try out your code as I would prefer to use the default close button rather than my custom close button.

The last question is in regard to tabbing out of the open panel and when/where to return the focus to the open panel.  If the user tabs from the action-bar item that is used to open the panel, should it set focus to the open panel?  Or instead of tab if the user presses the right arrow key from the action-bar item, should it set focus on the open Panel?

**Comment by macandcheese** [comment]
Going to cc @geospatialem for accessibility expectations for the last question - good topics.

**Comment by msamwill** [comment]
Hello,

Thank you for the reply.  My only concern with using the tab key to navigate from the action bar to the open panel is that the user has no way to tab directly to the next action bar item without closing the open panel.


________________________________
From: Kitty Hurley ***@***.***>
Sent: Wednesday, August 30, 2023 3:23 PM
To: Esri/calcite-design-system ***@***.***>
Cc: Williams, Anna ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL]: Re: [Esri/calcite-design-system] New Utility: focus manager + ShellPanel enhancement (#1471)


The last question is in regard to tabbing out of the open panel and when/where to return the focus to the open panel. If the user tabs from the action-bar item that is used to open the panel, should it set focus to the open panel? Or instead of tab if the user presses the right arrow key from the action-bar item, should it set focus on the open Panel?

If the user opens the panel from the action-bar, the panel's close button should be focused. This ensures context is provided to assistive technology (AT) users when an action has been performed to provide context.

If the panel is open by default, the action-bar's actions will follow navigation so the user can navigate through the available actions. AT reads back if the action is pressed or unpressed for context. Once navigation is complete on the action-bar, the panel context is provided.

As for tab vs arrow keys with the intent and elements of the components, tab and shift + tab are the recommended keys for consistent patterns with AT.

**Comment by geospatialem** [comment]
> My only concern with using the tab key to navigate from the action bar to the open panel is that the user has no way to tab directly to the next action bar item without closing the open panel.

Calcite recently introduced some navigation components, such as [`menu`](https://developers.arcgis.com/calcite-design-system/components/menu) and [`menu-item`](https://developers.arcgis.com/calcite-design-system/components/menu-item) which might accomplish what your aiming where the user could navigate through the menu with different keys. You'd need to setup the content to provide context to AT with the `menu-item`'s [`calciteMenuItemSelect`](https://developers.arcgis.com/calcite-design-system/components/menu-item/#api-reference-events-calciteMenuItemSelect) event with `aria-controls` and `aria-live`, something like https://codepen.io/geospatialem/pen/yLGOGGm.

**Comment by msamwill** [comment]
Thank you!  I see some really good examples in your code pens like these A11yDemos - a Collection by Kelly Hutchins on CodePen<https://codepen.io/collection/WvMEJM> and this one Calcite Maps SDK Search Widget Panel (codepen.io)<https://codepen.io/geospatialem/pen/oNJbvJR>.   The last one solves an issue I've been fighting with for quite some time getting the search widget to work correctly in a panel.

________________________________
From: Kitty Hurley ***@***.***>
Sent: Wednesday, August 30, 2023 4:31 PM
To: Esri/calcite-design-system ***@***.***>
Cc: Williams, Anna ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL]: Re: [Esri/calcite-design-system] New Utility: focus manager + ShellPanel enhancement (#1471)


My only concern with using the tab key to navigate from the action bar to the open panel is that the user has no way to tab directly to the next action bar item without closing the open panel.

Calcite recently introduced some navigation components, such as menu<https://urldefense.proofpoint.com/v2/url?u=https-3A__developers.arcgis.com_calcite-2Ddesign-2Dsystem_components_menu&d=DwMFaQ&c=j9YNQVTYIohH8cYE_3mLmw&r=PaO4_Cax22BUWnXzINx4sDIBHr4tHn5X2SQTEimO2dw&m=i33khJxCAQT2ca8zoj-n0FBwnz7MmmNp1oWO8wSmUx_7Jei_rDwx09DfcM_hlcyN&s=QOujPmOnhjfhG913Aaq_tpTxrAyPp_vksh_xZhZ5VgM&e=> and menu-item<https://urldefense.proofpoint.com/v2/url?u=https-3A__developers.arcgis.com_calcite-2Ddesign-2Dsystem_components_menu-2Ditem&d=DwMFaQ&c=j9YNQVTYIohH8cYE_3mLmw&r=PaO4_Cax22BUWnXzINx4sDIBHr4tHn5X2SQTEimO2dw&m=i33khJxCAQT2ca8zoj-n0FBwnz7MmmNp1oWO8wSmUx_7Jei_rDwx09DfcM_hlcyN&s=vJ5mZzJGvkOJ15RAAB0GLANXUuaBTgR6KpnBoDKdVNA&e=> which might accomplish what your aiming where the user could navigate through the menu with different keys. You'd need to setup the content to provide context to AT with the menu-item's calciteMenuItemSelect<https://urldefense.proofpoint.com/v2/url?u=https-3A__developers.arcgis.com_calcite-2Ddesign-2Dsystem_components_menu-2Ditem_-23api-2Dreference-2Devents-2DcalciteMenuItemSelect&d=DwMFaQ&c=j9YNQVTYIohH8cYE_3mLmw&r=PaO4_Cax22BUWnXzINx4sDIBHr4tHn5X2SQTEimO2dw&m=i33khJxCAQT2ca8zoj-n0FBwnz7MmmNp1oWO8wSmUx_7Jei_rDwx09DfcM_hlcyN&s=9wvLHHSmAos7aUpdQn5_pDQbZofDtNCMbIRrcdl1HmY&e=> event with aria-controls and aria-live, something like https://codepen.io/geospatialem/pen/yLGOGGm<https://urldefense.proofpoint.com/v2/url?u=https-3A__codepen.io_geospatialem_pen_yLGOGGm&d=DwMFaQ&c=j9YNQVTYIohH8cYE_3mLmw&r=PaO4_Cax22BUWnXzINx4sDIBHr4tHn5X2SQTEimO2dw&m=i33khJxCAQT2ca8zoj-n0FBwnz7MmmNp1oWO8wSmUx_7Jei_rDwx09DfcM_hlcyN&s=XiYgED-oGYoktgn6rgxMCEWjgzXmeQ0GIgaSNqQ6ZVA&e=>.

**Comment by msamwill** [comment]
Hello,

After reviewing demos, I found a lot of good content but not a best practice for keyboard navigation with calcite panels.  Based on what I've read in the posts I created this code pen Calcite panel keyboard navigation (codepen.io)<https://codepen.io/msamwill/pen/WNLxbbM>.  Can you tell me if this seems acceptable per A11y guidelines?

Codepen Description:
When a calcite panel is opened (made visible) focus is set on the panel.  When the panel is closed, focus returns to the corresponding action bar item that was used to open it.  If the user opens a panel, then uses shift tab to return focus to the action bar, they can use the right arrow key from the corresponding action bar item to return focus to the open panel.  Otherwise, they must close and reopen the panel to access it again from the action bar.

Thanks,

**Comment by geospatialem** [comment]
> After reviewing demos, I found a lot of good content but not a best practice for keyboard navigation with calcite panels. Based on what I've read in the posts I created this code pen Calcite panel keyboard navigation (codepen.io)<https://codepen.io/msamwill/pen/WNLxbbM>. Can you tell me if this seems acceptable per A11y guidelines?

@msamwill This follows the recommended pattern, but strongly encourage testing with larger more diverse audiences to continuously improve and support more folks. As another resource, Calcite also provides an [accessibility](https://developers.arcgis.com/calcite-design-system/foundations/accessibility/) page to consider designing for individuals and supporting accessibility in your apps.
